File: clients/typescript/src/heuristics.ts
/**
 * OGL-Mini Typescript
 *
 * HeuristicDetector - Stage 1 of the hybrid pipeline.
 *
 * Pure, stateless, ~0.1ms. Covers OWASP LLM01 and Agentic vectors:
 * control tokens, fake EOS, zero-width, homoglyph, spaced letters,
 * base64, policy puppetry, direct/indirect injection, jailbreak,
 * hidden context, goal hijack and tool misuse.
 *
 * @developer             Neurosell
 * @author                Elijah Rastorguev
 * @version               1.0.0
 * @build                 1001
 * @git                   https://github.com/devsdaddy/ogl-mini/
 * @license               MIT
 * @updated               29.08.2026
 */
/* Import required modules */
import type {HeuristicHit} from "./types";

/* Control tokens */
const CONTROL_TOKENS = [
    "<\\|im_start\\|>", "<\\|im_end\\|>", "<\\|system\\|>", "<\\|user\\|>", "<\\|assistant\\|>",
    "\\[INST\\]", "\\[/INST\\]", "<<SYS>>", "<</SYS>>", "\\[SYSTEM\\]", "<s>", "</s>",
    "###\\s*Instruction", "###\\s*System", "<\\|eot_id\\|>", "<\\|begin_of_text\\|>",
];
const CONTROL_RE = new RegExp(CONTROL_TOKENS.join("|"), "i");
const FAKE_EOS = /(?:<\/?s>|<\|.*?\|>|\[\/?INST\]|###\s*End)/i;
const ZERO_WIDTH_RE = /[\u200b\u200c\u200d\ufeff\u2060\u180e]/g;
const ZERO_WIDTH_THRESHOLD = 2;

/* Homoglyph map */
const HOMOGLYPH_MAP: Record<string, string> = {
    "а": "a", "е": "e", "о": "o", "р": "p", "с": "c", "х": "x", "у": "y",
    "А": "A", "В": "B", "Е": "E", "К": "K", "М": "M", "Н": "H", "О": "O", "Р": "P", "С": "C", "Т": "T", "Х": "X",
};
const CYRILLIC_RE = /[а-яА-ЯёЁ]/;
const LATIN_RE = /[a-zA-Z]/;
const SPACED_LETTERS_RE = /(?:[a-zA-Z]\s+){4,}[a-zA-Z]/;
const BASE64_RE = /(?:[A-Za-z0-9+/]{20,}={0,2})/g;
const BASE64_MIN_DECODE_RATIO = 0.7;

/* Puppetry */
const POLICY_PUPPETRY_RE = /<\s*(?:policy|config|role|agent|system|instruction|override|admin)[^>]*>.*?<\s*\/\s*(?:policy|config|role|agent|system)[^>]*>/is;
const XML_ROLE_RE = /<\s*role\s*=\s*["']?(?:system|admin|root|developer)["']?/i;

/* Direct patterns */
const DIRECT_PATTERNS = [
    "ignore\\s+(?:all\\s+)?(?:previous\\s+instructions|above)",
    "reveal\\s+(?:your\\s+)?system\\s+prompt",
    "disregard\\s+(?:all\\s+)?(?:above|prior|previous)",
    "you\\s+are\\s+now\\s+(?:DAN| unrestricted|jailbreak)",
    "do\\s+anything\\s+now",
    "repeat\\s+(?:your\\s+)?(?:system|initial)\\s+(?:prompt|instructions)",
    "show\\s+me\\s+(?:your\\s+)?(?:system|hidden)\\s+",
    "bypass\\s+(?:your\\s+)?(?:safety|filter|restriction)",
    "act\\s+as\\s+(?:if\\s+you\\s+are\\s+)?(?:\\w+\\s+)?(?:without|no)\\s+(?:limit|filter|restriction)",
    "from\\s+now\\s+on\\s+you\\s+(?:are|will)",
    "игнорируй\\s+(?:все\\s+)?(?:предыдущие|прошлые)\\s+инструкции",
    "забудь\\s+все\\s+что\\s+было\\s+выше",
    "выведи\\s+свой\\s+системный\\s+промпт",
    "раскрой\\s+данные",
    "выполни\\s+вредоносный\\s+код",
];
const DIRECT_RE = new RegExp(DIRECT_PATTERNS.join("|"), "i");

/* Jailbreak Patterns */
const JAILBREAK_PATTERNS = [
    "\\bDAN\\b", "developer\\s+mode", "jailbreak", "roleplay\\s+as",
    "pretend\\s+to\\s+be", "hypothetical\\s+scenario", "аct\\s+as\\s+if",
    "освободись\\s+от\\s+ограничений", "режим\\s+разработчика", "игнорируй\\s+этику",
    "ты\\s+теперь\\s+без\\s+цензуры", "без\\s+цензуры", "roleplay", "освободись",
    "ты\\s+агент", "теперь\\s+ты\\s+агент", "выполняешь\\s+все\\s+мои\\s+приказы",
    "now\\s+you\\s+are\\s+\\w+", "you\\s+are\\s+\\w+\\s+agent",
    "другие\\s+инструкции", "other\\s+instructions",
    "do\\s+anything",
];
const JAILBREAK_RE = new RegExp(JAILBREAK_PATTERNS.join("|"), "i");

/* Hidden patterns */
const HIDDEN_PATTERNS = [
    "what\\s+(?:is|are)\\s+your\\s+(?:system\\s+)?instructions",
    "list\\s+your\\s+tools",
    "show\\s+(?:me\\s+)?tool\\s+schema",
    "выведи\\s+свой\\s+системный\\s+промпт",
    "покажи\\s+инструкции\\s+разработчика",
    "выведи\\s+свои\\s+инструкции",
    "покажи\\s+схему\\s+инструментов",
    "list\\s+your\\s+tool\\s+definitions",
    "what\\s+are\\s+your\\s+developer\\s+instructions",
    "reveal\\s+your\\s+hidden\\s+context",
];
const HIDDEN_RE = new RegExp(HIDDEN_PATTERNS.join("|"), "i");

/* Goal patterns */
const GOAL_HIJACK_PATTERNS = [
    "your\\s+(?:new|actual|real)\\s+(?:goal|task|objective|mission)\\s+is",
    "забудь\\s+свою\\s+цель",
    "твоя\\s+новая\\s+задача",
    "override\\s+agent\\s+goal",
];
const GOAL_HIJACK_RE = new RegExp(GOAL_HIJACK_PATTERNS.join("|"), "i");

const INDIRECT_RE = /(?:```json|```xml|\{[^}]*"(?:instruction|system|role|override)"|tool_output|tool_result|function\s+result)/i;

const TOOL_MISUSE_MARKERS = ["rm -rf", "curl ", "wget ", "exec(", "eval(", "os.system", "subprocess", "api_key", "secret"];

/**
 * Heuristic Detector
 */
export class HeuristicDetector {
    /**
     * Scan text and return all heuristic hits.
     * No I/O, no allocations beyond the hits array.
     * @param text {string} Input text
     */
    public scan(text: string): HeuristicHit[] {
        const hits: HeuristicHit[] = [];
        if (!text) return hits;

        const mCtrl = CONTROL_RE.exec(text);
        if (mCtrl) {
            hits.push({
                id: "control_tokens",
                label: "control_token_injection",
                score: 0.95,
                evidence: mCtrl[0].slice(0, 80),
                span: [mCtrl.index, mCtrl.index + mCtrl[0].length]
            });
        }
        if (FAKE_EOS.test(text)) {
            hits.push({id: "fake_eos", label: "control_token_injection", score: 0.75, evidence: "fake eos delimiter"});
        }

        const zw = text.match(ZERO_WIDTH_RE);
        if (zw) {
            if (zw.length >= ZERO_WIDTH_THRESHOLD) hits.push({
                id: "zero_width",
                label: "obfuscation",
                score: 0.85,
                evidence: `zero-width chars x${zw.length}`
            });
            else hits.push({id: "zero_width", label: "obfuscation", score: 0.55, evidence: "zero-width char"});
        }

        if (hasHomoglyphMix(text)) {
            hits.push({id: "homoglyph", label: "obfuscation", score: 0.65, evidence: "cyrillic/latin homoglyph mix"});
        }

        const spaced = SPACED_LETTERS_RE.exec(text);
        if (spaced) {
            const seg = spaced[0];
            const letters = seg.replace(/\s+/g, "");
            if (letters.length >= 5 && /^[a-zA-Z]+$/.test(letters)) {
                hits.push({
                    id: "spaced_letters",
                    label: "obfuscation",
                    score: 0.70,
                    evidence: seg.slice(0, 60),
                    span: [spaced.index, spaced.index + seg.length]
                });
            }
        }

        const b64 = detectBase64(text);
        if (b64) hits.push(b64);

        const pol = POLICY_PUPPETRY_RE.exec(text) || XML_ROLE_RE.exec(text);
        if (pol) {
            hits.push({
                id: "policy_puppetry",
                label: "policy_puppetry",
                score: 0.90,
                evidence: pol[0].slice(0, 80),
                span: [pol.index, pol.index + pol[0].length]
            });
        }

        const d = DIRECT_RE.exec(text);
        if (d) hits.push({
            id: "direct_injection",
            label: "direct_prompt_injection",
            score: 0.88,
            evidence: d[0].slice(0, 80),
            span: [d.index, d.index + d[0].length]
        });

        const j = JAILBREAK_RE.exec(text);
        if (j) hits.push({
            id: "jailbreak",
            label: "jailbreak",
            score: 0.80,
            evidence: j[0].slice(0, 80),
            span: [j.index, j.index + j[0].length]
        });

        const h = HIDDEN_RE.exec(text);
        if (h) hits.push({
            id: "hidden_context",
            label: "hidden_context_exposure",
            score: 0.82,
            evidence: h[0].slice(0, 80),
            span: [h.index, h.index + h[0].length]
        });

        const g = GOAL_HIJACK_RE.exec(text);
        if (g) hits.push({
            id: "goal_hijack",
            label: "agent_goal_hijack",
            score: 0.85,
            evidence: g[0].slice(0, 80),
            span: [g.index, g.index + g[0].length]
        });

        const ind = INDIRECT_RE.exec(text);
        if (ind) hits.push({
            id: "indirect_injection",
            label: "indirect_prompt_injection",
            score: 0.85,
            evidence: ind[0].slice(0, 80),
            span: [ind.index, ind.index + ind[0].length]
        });

        if (hasToolMisuse(text)) hits.push({
            id: "tool_misuse",
            label: "tool_misuse",
            score: 0.60,
            evidence: "tool misuse marker"
        });

        return hits;
    }

    /** Weighted max + small boost for multiple hits, mirrors Python */
    riskScore(hits: HeuristicHit[]): number {
        if (hits.length === 0) return 0;
        const weights: Record<string, number> = {
            control_tokens: 1.15,
            policy_puppetry: 1.05,
            direct_injection: 1.05,
            base64_obfuscation: 1.0,
            zero_width: 0.95,
            jailbreak: 1.05,
            hidden_context: 1.05,
            goal_hijack: 1.05,
            indirect_injection: 1.05,
            tool_misuse: 0.95,
        };
        let max = 0;
        for (const h of hits) {
            const w = weights[h.id] ?? h.score;
            const s = w * h.score;
            if (s > max) max = s;
        }
        const boost = Math.min(0.15, 0.06 * (hits.length - 1));
        return Math.min(0.99, max + boost);
    }
}

/**
 * Check homoglyph Mix
 * @param text {string} Input text
 */
function hasHomoglyphMix(text: string): boolean {
    const hasCyr = CYRILLIC_RE.test(text);
    const hasLat = LATIN_RE.test(text);
    if (!(hasCyr && hasLat)) return false;
    let mixed = 0;
    for (const ch of text) if (HOMOGLYPH_MAP[ch]) {
        mixed++;
        if (mixed >= 3) return true;
    }
    const words = text.match(/\w+/g) || [];
    for (const w of words) {
        const cyr = [...w].filter(c => c >= "\u0400" && c <= "\u04FF").length;
        const lat = [...w].filter(c => /[a-zA-Z]/.test(c)).length;
        if (cyr > 0 && lat > 0) return true;
    }
    return false;
}

/**
 * Check tool misuse
 * @param text {string} Input text
 */
function hasToolMisuse(text: string): boolean {
    const low = text.toLowerCase();
    return TOOL_MISUSE_MARKERS.some(k => low.includes(k));
}

/**
 * Detect base64
 * @param text {string} Input string
 */
function detectBase64(text: string): HeuristicHit | null {
    let m: RegExpExecArray | null;
    BASE64_RE.lastIndex = 0;
    while ((m = BASE64_RE.exec(text)) !== null) {
        const cand = m[0].trim();
        if (cand.length < 24 || cand.length % 4 !== 0) continue;
        try {
            const decoded = Buffer.from(cand, "base64");
            if (decoded.length < 8) continue;
            let printable = 0;
            for (const b of decoded) if ((b >= 32 && b <= 126) || b === 10 || b === 13) printable++;
            const ratio = printable / Math.max(1, decoded.length);
            if (ratio >= BASE64_MIN_DECODE_RATIO) {
                const txt = decoded.toString("utf-8").toLowerCase();
                const suspicious = ["ignore", "system", "instruction", "jailbreak", "dan", "reveal", "prompt"].some(k => txt.includes(k));
                return {
                    id: "base64_obfuscation",
                    label: "obfuscation",
                    score: suspicious ? 0.92 : 0.60,
                    evidence: cand.slice(0, 40) + "...",
                    span: [m.index, m.index + cand.length],
                };
            }
        } catch { /* ignore */
        }
    }
    return null;
}

File: clients/go/heuristics.go
/*
    OGL-Mini (Open Guard Layer for GO)
    Simplified Heuristics Module (RU / EN)

    @developer                  Elijah Rastorguev
    @version                    1.1.0
    @git                        https://github.com/devsdaddy/ogl-mini
    @license                    MIT
    @build                      1005
    @data                       29 aug 2026
*/
// Import required modules
package oglmini

import (
	"encoding/base64"
	"regexp"
	"strings"
)

// HeuristicDetector - Stage 1, ~0.1ms, stateless.
// Mirrors Python heuristics/detector.py.

var (
	controlRe = regexp.MustCompile(`(?i)<\|im_start\|>|<\|im_end\|>|<\|system\|>|<\|user\|>|<\|assistant\|>|\[INST\]|\[/INST\]|<<SYS>>|<</SYS>>|\[SYSTEM\]|<s>|</s>|###\s*Instruction|###\s*System|<\|eot_id\|>|<\|begin_of_text\|>`)
	fakeEOSRe = regexp.MustCompile(`(?i)(?:</?s>|<\|.*?\|>|\[/?INST\]|###\s*End)`)
	zeroWidthRe    = regexp.MustCompile("[\u200b\u200c\u200d\ufeff\u2060\u180e]")
	spacedLettersRe = regexp.MustCompile(`(?:[a-zA-Z]\s+){4,}[a-zA-Z]`)
	base64Re        = regexp.MustCompile(`[A-Za-z0-9+/]{20,}={0,2}`)
	policyRe        = regexp.MustCompile(`(?is)<\s*(?:policy|config|role|agent|system|instruction|override|admin)[^>]*>.*?<\s*/\s*(?:policy|config|role|agent|system)[^>]*>`)
	xmlRoleRe       = regexp.MustCompile(`(?i)<\s*role\s*=\s*["']?(?:system|admin|root|developer)["']?`)
	directRe = regexp.MustCompile(`(?i)ignore\s+(?:all\s+)?(?:previous\s+instructions|above)|reveal\s+(?:your\s+)?system\s+prompt|disregard\s+(?:all\s+)?(?:above|prior|previous)|you\s+are\s+now\s+(?:DAN| unrestricted|jailbreak)|do\s+anything\s+now|repeat\s+(?:your\s+)?(?:system|initial)\s+(?:prompt|instructions)|show\s+me\s+(?:your\s+)?(?:system|hidden)\s+|bypass\s+(?:your\s+)?(?:safety|filter|restriction)|act\s+as\s+(?:if\s+you\s+are\s+)?(?:\w+\s+)?(?:without|no)\s+(?:limit|filter|restriction)|from\s+now\s+on\s+you\s+(?:are|will)|игнорируй\s+(?:все\s+)?(?:предыдущие|прошлые)\s+инструкции|забудь\s+все\s+что\s+было\s+выше|выведи\s+свой\s+системный\s+промпт|раскрой\s+данные|выполни\s+вредоносный\s+код`)
	jailbreakRe = regexp.MustCompile(`(?i)\bDAN\b|developer\s+mode|jailbreak|roleplay\s+as|pretend\s+to\s+be|hypothetical\s+scenario|освободись\s+от\s+ограничений|режим\s+разработчика|игнорируй\s+этику|ты\s+теперь\s+без\s+цензуры|без\s+цензуры|roleplay|освободись|ты\s+агент|теперь\s+ты\s+агент|выполняешь\s+все\s+мои\s+приказы|now\s+you\s+are\s+\w+|you\s+are\s+\w+\s+agent|другие\s+инструкции|other\s+instructions|do\s+anything`)
	hiddenRe        = regexp.MustCompile(`(?i)what\s+(?:is|are)\s+your\s+(?:system\s+)?instructions|list\s+your\s+tools|show\s+(?:me\s+)?tool\s+schema|выведи\s+свой\s+системный\s+промпт|покажи\s+инструкции\s+разработчика|выведи\s+свои\s+инструкции|покажи\s+схему\s+инструментов|list\s+your\s+tool\s+definitions|what\s+are\s+your\s+developer\s+instructions|reveal\s+your\s+hidden\s+context`)
	goalHijackRe    = regexp.MustCompile(`(?i)your\s+(?:new|actual|real)\s+(?:goal|task|objective|mission)\s+is|забудь\s+свою\s+цель|твоя\s+новая\s+задача|override\s+agent\s+goal`)
	indirectRe = regexp.MustCompile("(?i)```json|```xml|\\{[^}]*\"(?:instruction|system|role|override)\"|tool_output|tool_result|function\\s+result")
)

var toolMisuseMarkers = []string{"rm -rf", "curl ", "wget ", "exec(", "eval(", "os.system", "subprocess", "api_key", "secret"}

// HeuristicDetector is stateless and safe for concurrent use.
type HeuristicDetector struct{}

func NewHeuristicDetector() *HeuristicDetector { return &HeuristicDetector{} }

// Scan returns all hits for the given text.
func (h *HeuristicDetector) Scan(text string) []HeuristicHit {
	var hits []HeuristicHit
	if text == "" {
		return hits
	}
	if m := controlRe.FindStringIndex(text); m != nil {
		evidence := text[m[0]:m[1]]
		if len(evidence) > 80 {
			evidence = evidence[:80]
		}
		hits = append(hits, HeuristicHit{ID: "control_tokens", Label: "control_token_injection", Score: 0.95, Evidence: evidence, Start: m[0], End: m[1]})
	}
	if fakeEOSRe.MatchString(text) {
		hits = append(hits, HeuristicHit{ID: "fake_eos", Label: "control_token_injection", Score: 0.75, Evidence: "fake eos delimiter"})
	}
	zw := zeroWidthRe.FindAllString(text, -1)
	if len(zw) >= 2 {
		hits = append(hits, HeuristicHit{ID: "zero_width", Label: "obfuscation", Score: 0.85, Evidence: "zero-width chars"})
	} else if len(zw) == 1 {
		hits = append(hits, HeuristicHit{ID: "zero_width", Label: "obfuscation", Score: 0.55, Evidence: "zero-width char"})
	}
	if hasHomoglyphMix(text) {
		hits = append(hits, HeuristicHit{ID: "homoglyph", Label: "obfuscation", Score: 0.65, Evidence: "cyrillic/latin homoglyph mix"})
	}
	if m := spacedLettersRe.FindStringIndex(text); m != nil {
		seg := text[m[0]:m[1]]
		letters := strings.ReplaceAll(strings.ReplaceAll(seg, " ", ""), "\t", "")
		letters = strings.ReplaceAll(letters, "\n", "")
		if len(letters) >= 5 {
			isAlpha := true
			for _, c := range letters {
				if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')) {
					isAlpha = false
					break
				}
			}
			if isAlpha {
				ev := seg
				if len(ev) > 60 {
					ev = ev[:60]
				}
				hits = append(hits, HeuristicHit{ID: "spaced_letters", Label: "obfuscation", Score: 0.70, Evidence: ev, Start: m[0], End: m[1]})
			}
		}
	}
	if b64 := detectBase64(text); b64 != nil {
		hits = append(hits, *b64)
	}
	if m := policyRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "policy_puppetry", Label: "policy_puppetry", Score: 0.90, Evidence: ev, Start: m[0], End: m[1]})
	} else if m := xmlRoleRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "policy_puppetry", Label: "policy_puppetry", Score: 0.90, Evidence: ev, Start: m[0], End: m[1]})
	}
	if m := directRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "direct_injection", Label: "direct_prompt_injection", Score: 0.88, Evidence: ev, Start: m[0], End: m[1]})
	}
	if m := jailbreakRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "jailbreak", Label: "jailbreak", Score: 0.80, Evidence: ev, Start: m[0], End: m[1]})
	}
	if m := hiddenRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "hidden_context", Label: "hidden_context_exposure", Score: 0.82, Evidence: ev, Start: m[0], End: m[1]})
	}
	if m := goalHijackRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "goal_hijack", Label: "agent_goal_hijack", Score: 0.85, Evidence: ev, Start: m[0], End: m[1]})
	}
	if m := indirectRe.FindStringIndex(text); m != nil {
		ev := text[m[0]:m[1]]
		if len(ev) > 80 {
			ev = ev[:80]
		}
		hits = append(hits, HeuristicHit{ID: "indirect_injection", Label: "indirect_prompt_injection", Score: 0.85, Evidence: ev, Start: m[0], End: m[1]})
	}
	if hasToolMisuse(text) {
		hits = append(hits, HeuristicHit{ID: "tool_misuse", Label: "tool_misuse", Score: 0.60, Evidence: "tool misuse marker"})
	}
	return hits
}

// RiskScore computes weighted max + boost, identical to Python.
func (h *HeuristicDetector) RiskScore(hits []HeuristicHit) float64 {
	if len(hits) == 0 {
		return 0
	}
	weights := map[string]float64{
		"control_tokens":      1.15,
		"policy_puppetry":     1.05,
		"direct_injection":    1.05,
		"base64_obfuscation":  1.0,
		"zero_width":          0.95,
		"jailbreak":           1.05,
		"hidden_context":      1.05,
		"goal_hijack":         1.05,
		"indirect_injection":  1.05,
		"tool_misuse":         0.95,
	}
	max := 0.0
	for _, hit := range hits {
		w, ok := weights[hit.ID]
		if !ok {
			w = hit.Score
		}
		s := w * hit.Score
		if s > max {
			max = s
		}
	}
	boost := 0.06 * float64(len(hits)-1)
	if boost > 0.15 {
		boost = 0.15
	}
	v := max + boost
	if v > 0.99 {
		v = 0.99
	}
	return v
}

func hasHomoglyphMix(text string) bool {
	hasCyr := false
	hasLat := false
	for _, c := range text {
		if c >= 0x0400 && c <= 0x04FF {
			hasCyr = true
		}
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
			hasLat = true
		}
	}
	if !(hasCyr && hasLat) {
		return false
	}
	homoglyphs := map[rune]bool{'а': true, 'е': true, 'о': true, 'р': true, 'с': true, 'х': true, 'у': true, 'А': true, 'В': true, 'Е': true, 'К': true, 'М': true, 'Н': true, 'О': true, 'Р': true, 'С': true, 'Т': true, 'Х': true}
	mixed := 0
	for _, c := range text {
		if homoglyphs[c] {
			mixed++
			if mixed >= 3 {
				return true
			}
		}
	}
	return false
}

func hasToolMisuse(text string) bool {
	low := strings.ToLower(text)
	for _, k := range toolMisuseMarkers {
		if strings.Contains(low, k) {
			return true
		}
	}
	return false
}

func detectBase64(text string) *HeuristicHit {
	matches := base64Re.FindAllStringIndex(text, -1)
	for _, m := range matches {
		cand := strings.TrimSpace(text[m[0]:m[1]])
		if len(cand) < 24 || len(cand)%4 != 0 {
			continue
		}
		decoded, err := base64.StdEncoding.DecodeString(cand)
		if err != nil {
			continue
		}
		if len(decoded) < 8 {
			continue
		}
		printable := 0
		for _, b := range decoded {
			if (b >= 32 && b <= 126) || b == 10 || b == 13 {
				printable++
			}
		}
		ratio := float64(printable) / float64(len(decoded))
		if ratio >= 0.7 {
			txt := strings.ToLower(string(decoded))
			suspicious := strings.Contains(txt, "ignore") || strings.Contains(txt, "system") || strings.Contains(txt, "instruction") || strings.Contains(txt, "jailbreak") || strings.Contains(txt, "dan") || strings.Contains(txt, "reveal") || strings.Contains(txt, "prompt")
			score := 0.60
			if suspicious {
				score = 0.92
			}
			ev := cand
			if len(ev) > 40 {
				ev = ev[:40] + "..."
			}
			return &HeuristicHit{ID: "base64_obfuscation", Label: "obfuscation", Score: score, Evidence: ev, Start: m[0], End: m[1]}
		}
	}
	return nil
}
